compiler/hash-ir/src/ty.rs

```diff
@@ -0,0 +1,202 @@
+//! Simplified Hash type hierarchy. The [IrTy] is used a simplified
+//! variant of the available Hash [Term]s that are defined in the
+//! `hash-types` crate. This data structure is used to remove unnecessary
+//! complexity from types that are required for IR generation and
+//! analysis.
+
+use std::fmt;
+
+use bitflags::bitflags;
+use hash_source::{
+    constant::{FloatTy, SIntTy, UIntTy},
+    identifier::Identifier,
+};
+use hash_utils::{
+    new_sequence_store_key, new_store_key,
+    store::{CloneStore, DefaultSequenceStore, DefaultStore, SequenceStore, Store},
+};
+use index_vec::IndexVec;
+
+use crate::write::{ForFormatting, WriteTyIr};
+
+/// Simplified type structure used by the IR and other stages to reason about
+/// Hash programs once types have been erased and simplified.
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+pub enum IrTy {
+    /// A primitive signed integer type, e.g. `i32`.
+    Int(SIntTy),
+
+    /// A primitive unsigned integer type, e.g. `u32`.
+    UInt(UIntTy),
+
+    /// A primitive floating point type, e.g. `f32`.
+    Float(FloatTy),
+
+    /// String type.
+    Str,
+
+    /// Boolean type.
+    Bool,
+
+    /// Character type.
+    Char,
+
+    /// The never type
+    Never,
+
+    /// A reference type, referring to a another type, e.g. `&T`
+    Ref(IrTyId),
```

Comment:
What about different reference variants?

Reply:
The other variants are `RawRef` and ...?

Reply:
Mutable ones, local refs vs RC refs

Reply:
For local ones, do you mean ones that point to the stack? If yes, is there really a need to differentiate between `Ref` and `StackRef`?